### PR TITLE
[Merged by Bors] - refactor: use `List.splitAt` in `work_on_goal`

### PR DESCRIPTION
**Pull request by arthurpaulino**
This implementation traverses the list of goals less times.

It also adds some functions that can be useful in the future (where should they go?)

### CODE REVIEW

File: Mathlib/Tactic/Basic.lean [hunk 1]
```diff
@@ -244,17 +244,23 @@ elab "any_goals " seq:tacticSeq : tactic => do
     throwError "failed on all goals"
   setGoals mvarIdsNew.toList
 
+-- TODO: move these
+def List.splitAux {α : Type} : List α → Nat → List α → List α × List α
+  | h :: t, n + 1, acc => splitAux t n (h :: acc)
+  | l,      _,     acc => (acc.reverse, l)
+
+def List.split {α : Type} (l : List α) (i : Nat) : List α × List α :=
```
Comment (gebner) [inline Mathlib/Tactic/Basic.lean:252]: I believe this function is called `List.splitAt` and is already in mathlib4.

**Review by gebner — COMMENTED** [review]
"Optimizing" is maybe overselling the change a bit. :smile:   The list of goals is typically single-digit.  But it's still a good change that makes the code more readable.

Please also adhere to the PR title convention.  I would rename the PR to something like `refactor: use List.splitAt in work_on_goal`.

**Comment by gebner** [comment]
You need to put the new file into `Mathlib.lean`.  Otherwise lgtm.

bors d+

File: Mathlib/Tactic/Basic.lean [hunk 2]
```diff
@@ -243,18 +243,3 @@ elab "any_goals " seq:tacticSeq : tactic => do
   if not anySuccess then
     throwError "failed on all goals"
   setGoals mvarIdsNew.toList
-
-/--
-`work_on_goal n { tac }` creates a block scope for the `n`-th goal (indexed from zero),
```
Comment (arthurpaulino) [inline Mathlib/Tactic/Basic.lean:248]: Is this syntax accurate or should I change it too?

**Pull request closed without merging**